### ISSUE TITLE
cy.getCookie() does not support timeouts

### ISSUE THREAD

**Issue by webnexusmobile** [cypress-io/cypress#4802]
### Current behavior:
Both of the following lines result in 'expected null to exist' although the cookie becomes available within about a second: 
cy.getCookie('token').should('exist')
cy.getCookie('token', { timeout: 5000 }).should('exist')

### Desired behavior:
cy.getCookie should use the default timeout when used together with 'should'
cy.getCookie should support the timeout argument as documented: https://docs.cypress.io/api/commands/getcookie.html#Arguments

### Steps to reproduce: (app code and test code)
Use cy.getCookie('someCookie').should('exist') for a cookie which is saved with a little delay, e.g. session cookie after login.

### Versions
cypress 3.4.0, electron 61, 5.2.1-arch1-1-ARCH

**Comment by jennifer-shehane** [comment]
`cy.getCookie()` looks to use the default [`responseTimeout`](https://on.cypress.io/configuration#Timeouts) of 30seconds.

code: https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cy/commands/cookies.coffee#L80

We do have some tests that this timeout is being sent through correctly to the command - and that it times out if it is not present in this time, but we actually do *not* have a test for called `cy.getCookie()` with the cookie initially not existing, setting the cookie, and ensuring the test passes. 

I'll go ahead and add a test for this in our codebase to see if this is working.

**Comment by jainmitesh09** [comment]
Hi @jennifer-shehane,

I dont think 30 seconds timeout exists. Any particular version to look out for? 

Also, passing a timeout param would be great! 

Thanks!

**Comment by abhidp** [comment]
I too faced the same issue. Moving the `cy.getCookie()` call to below the `cy.url()` call solved the issue as `cy.url()` automatically waits for default timeout. By this time, the cookie would have already been issued. More details below:

```javascript
// THIS CODE FAILED 👎 
it('should login successfully with correct credentials', () => {
    cy.login();

   // BECAUSE CYPRESS DID NOT WAIT FOR THE COOKIE TO BE ISSUED
    cy.getCookie('auth').should('exist');
    cy.title().should('eq', 'Crates for everyone! - Crate');
    cy.url().should('eq', `${Cypress.config().baseUrl}/crates`);
 })
```

```javascript
// THIS CODE PASSED AFTER MOVING THE getCookie() call to the bottom👍 
it('should login successfully with correct credentials', () => {
    cy.login()

   // BECAUSE CYPRESS WAITED FOR THE url() and title() CALLS, AND BY THAT TIME 
   // THE COOKIE WOULD HAVE OBVIOUSLY BEEN ISSUED
    cy.title().should('eq', 'Crates for everyone! - Crate');
    cy.url().should('eq', `${Cypress.config().baseUrl}/crates`)
    cy.getCookie('auth').should('exist');
 })
```

**Comment by acurrieebix** [comment]
I too face this issue.  There is a timeout option according to the documentation but this seems to be ignored on testing.  I used the simple recipe of html forms login and I can see that the cookie issues so quickly which is why it passes.  It real world with slight delay, it does not and an error "expected null to exist" is encountered.

Another issue which has been open for a whole year without resolution.  This is very disappointing.

**Comment by ahmed1293** [comment]
Experiencing exactly the same thing after logging in and trying to `get` an auth cookie. Adding a `wait` fixes the issue, but isn't ideal.

**Comment by mohsenny** [comment]
I have a similar issue. I'm passing a timeout of 10 seconds, but the test immediately faces a `expects null to exists` in this line and without respecting the `timeout: 1000`
```
...
.getCookie('SESSIONID', { timeout: 10000 }).should('exist')
...
```

**Comment by MalteJoe** [comment]
I too faced this issue. The docs (and TypeScript types) are misleading as `timeout` does nothing if a command isn't retried anyway right (according to https://docs.cypress.io/api/commands/getcookie.html#Assertions)? The cookie I'm looking for is sent from the backend and is needed for authorized requests which I need for my test setup.

**Comment by johot** [comment]
Same here :( this seems like such a basic functionality hope it can get fixed soon! Thank you Cypress :)

**Comment by lucksp** [comment]
```
cy.getCookie('user', {
      timeout: 30000,
    })
```

using:
```
"@testing-library/cypress@^7.0.3":
cypress@^6.6.1:
```

I too seem to have the same issue where the `timeout` is appearing to be ignored:
![image](https://user-images.githubusercontent.com/15162169/109569358-8edd6a80-7aa5-11eb-96f0-61ee74484714.png)

**Comment by theKashey** [comment]
A little better workaround is to use https://github.com/NoriSte/cypress-wait-until
```ts
// wait until a cookie is set
cy.waitUntil(() => cy.getCookie('token').then(cookie => Boolean(cookie && cookie.value)));
```

**Comment by tgamauf** [comment]
> A little better workaround is to use https://github.com/NoriSte/cypress-wait-until
> 
> ```ts
> // wait until a cookie is set
> cy.waitUntil(() => cy.getCookie('token').then(cookie => Boolean(cookie && cookie.value)));
> ```

Doesn't fix the issue for me.

**Comment by stokrattt** [comment]
Doesn't fix the issue for me.

**Comment by theKashey** [comment]
How exactly it’s not working?

**Comment by stokrattt** [comment]
After API login in before, the cookie (in before is too) is null (empty) on the CI. Locally is normal.

**Comment by intellix** [comment]
This documentation is wrong and should be fixed at least unil this works as expected: https://docs.cypress.io/api/commands/getcookie#Session-id

"expected null to exist"

**Comment by liambutler** [comment]
Recording of this issue in action:
https://user-images.githubusercontent.com/24389501/168602990-48e7f37f-f25d-487b-859c-ef3bf07d9ae9.mp4

According to the [docs](https://docs.cypress.io/api/commands/should#Function) for `.should()`:

> Passing a function to .should() enables you to make multiple assertions on the yielded subject. 

However you can see from the video that this fails on the first assertion, and does not retry

**Comment by wilsonpage** [comment]
> .should('exist')

Yeah this works for me. The code i'm using is:

```ts
cy.waitUntil(() =>
  cy
    .getCookie(CLIENT_ID_COOKIE_NAME, { log: false })
    .then((result) => result?.value)
  ).then((cookieValue) => {
    // do something with cookieValue
});
```

It retries until the cookie is set.

**Comment by emilyrohrbough** [comment]
The issue here is `getCookie()` is resolving within the timeout, but it will not retry assertions (or re-querying for the expected value) until the timeout has expired. The `cy.getCookie()` command needs to be updated to become a [query command](https://docs.cypress.io/guides/core-concepts/retry-ability#Commands-Queries-and-Assertions) so this behaves as expected by re-pulling the value and running the assertions without workaround suggested above.

**Comment by andrei9669** [comment]
just stumbled upon this issue, I guess wait(500) will work meanwhile

**Comment by WillsB3** [comment]
> The issue here is `getCookie()` is resolving within the timeout, but it will not retry assertions (or re-querying for the expected value) until the timeout has expired. The `cy.getCookie()` command needs to be updated to become a [query command](https://docs.cypress.io/guides/core-concepts/retry-ability#Commands-Queries-and-Assertions) so this behaves as expected by re-pulling the value and running the assertions without workaround suggested above.

I came to the same conclusion, however at the moment I think getCookie in Cypress is Async. If this is the case does it mean we’ll never see a query version of getCookie?

**Comment by variable** [comment]
Experiencing the same problem.

**Comment by DaydreamsDev** [comment]
Issue has been routed to backlog. This is not a commitment to have it prioritized by the team.  :confused:

**Comment by jennifer-shehane** [comment]
This will be included as part of our upcoming Cypress 16 release. The PR is already merged on that branch. 

It is considered breaking since it will change to a query type command, which means if anyone overwrote it - they would need to change their signature. 

We’ll comment on the issue once Cypress 16 is ready to install.
